clients/src/main/java/org/apache/kafka/clients/admin/ElectPreferredLeadersResult.java

```diff
@@ -0,0 +1,130 @@
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The result of {@link AdminClient#electPreferredLeaders(Collection, ElectPreferredLeadersOptions)}
+ *
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class ElectPreferredLeadersResult {
+
+    private final KafkaFutureImpl<Map<TopicPartition, ApiError>> electionFuture;
+    private final Set<TopicPartition> partitions;
+
+    ElectPreferredLeadersResult(KafkaFutureImpl<Map<TopicPartition, ApiError>> electionFuture, Set<TopicPartition> partitions) {
+        this.electionFuture = electionFuture;
+        this.partitions = partitions;
+    }
+
+    /**
+     * Get the result of the election for the given {@code partition}.
+     * If there was not an election triggered for the given {@code partition}, the
+     * returned future will complete with an error.
+     */
+    public KafkaFuture<Void> partitionResult(final TopicPartition partition) {
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+        electionFuture.addWaiter(new KafkaFuture.BiConsumer<Map<TopicPartition, ApiError>, Throwable>() {
+            @Override
+            public void accept(Map<TopicPartition, ApiError> topicPartitions, Throwable throwable) {
+                if (throwable != null) {
+                    result.completeExceptionally(throwable);
+                } else if (!topicPartitions.containsKey(partition)) {
+                    result.completeExceptionally(new UnknownTopicOrPartitionException(
+                            "Preferred leader election for partition \"" + partition +
+                                    "\" was not attempted"));
+                } else {
+                    ApiException exception = topicPartitions.get(partition).exception();
+                    if (exception == null) {
+                        result.complete(null);
+                    } else {
+                        result.completeExceptionally(exception);
+                    }
+                }
+            }
+        });
+        return result;
+    }
+
+    /**
+     * <p>Get a future for the topic partitions for which a leader election
+     * was attempted. A partition will be present in this result if
+     * an election was attempted even if the election was not successful.</p>
+     *
+     * <p>This method is provided to discover the partitions attempted when
+     * {@link AdminClient#electPreferredLeaders(Collection)} is called
+     * with a null {@code partitions} argument.</p>
+     */
+    public KafkaFuture<Set<TopicPartition>> partitions() {
+        if (partitions != null) {
+            return KafkaFutureImpl.completedFuture(this.partitions);
+        } else {
+            final KafkaFutureImpl<Set<TopicPartition>> result = new KafkaFutureImpl<>();
+            electionFuture.addWaiter(new KafkaFuture.BiConsumer<Map<TopicPartition, ApiError>, Throwable>() {
```

Comment:
This really should be `whenComplete` rather than `addWaiter`.  `addWaiter` is an internal API which is only supposed to be used inside the KafkaFuture code...